fem/spacehierarchy.hpp

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_SPACEHIERARCHY
+#define MFEM_SPACEHIERARCHY
+
+#include "fespace.hpp"
+
+#ifdef MFEM_USE_MPI
+#include "pfespace.hpp"
+#endif
+
+namespace mfem
+{
+
+/// Class bundling a hierarchy of meshes and finite element spaces
+template <typename M, typename FES> class AbstractSpaceHierarchy
```

Comment:
This was my initial idea idea, but it makes the class much more complicated for users to use. For instance, if a user wants to create a new ParBilinearForm using a FiniteElementSpace returned by the SpaceHierarchy, he needs to cast it manually to a ParFiniteElementSpace. Another issue arises in the AddUniformlyRefinedLevel method where a new FiniteElementSpace is being allocated and the required type can (for my understanding) not be deducted at run time.

Reply:
One way of checking the type is performing a `dynamic_cast` to e.g. `ParFiniteElementSpace *` and checking for `nullptr`

Reply:
In addition to using Will's suggestion in `AddUniformlyRefinedLevel`, I suggest removing all the GetMesh methods (the user can do `GetFESpace().GetMesh()`), and have a `ParSpaceHierarchy` object that overloads only the GetFESpace() type methods.

There are other options, but I think this would fit into existing MFEM code the best. We tend to avoid templates (whether or not that's a good decision, but it's a decision that's already been made).

Reply:
I rewrote the SpaceHierarchy using polymorphism and overloading so it fits better into the current MFEM code.